linera-core/src/client/mod.rs

```diff
@@ -1144,73 +1146,105 @@ where
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
+        let remote_node_chains_view = info.requested_received_log.into_iter().fold(
+            BTreeMap::<ChainId, Vec<ChainAndHeight>>::new(),
+            |mut chain_to_info, entry| {
+                chain_to_info.entry(entry.chain_id).or_default().push(entry);
+                chain_to_info
+            },
+        );
+
         let mut certificates: Vec<Certificate> = Vec::new();
         let mut new_tracker = tracker;
 
-        for entry in info.requested_received_log {
-            let query = ChainInfoQuery::new(entry.chain_id)
-                .with_sent_certificate_hashes_in_range(BlockHeightRange::single(entry.height));
+        for (chain_id, blocks) in remote_node_chains_view {
+            debug_assert!({
+                blocks.iter().fold(BlockHeight::ZERO, |prev, curr| {
+                    if curr.height < prev {
+                        panic!("Received blocks out of order");
+                    }
+                    curr.height
+                });
+                true
+            });
 
-            let local_response = self
-                .client
-                .local_node
-                .handle_chain_info_query(query.clone())
-                .await
-                .map_err(|error| NodeError::LocalNodeQuery {
-                    error: error.to_string(),
-                })?
-                .info;
-            if !local_response.requested_sent_certificate_hashes.is_empty() {
-                // We've already processed incoming messages for this certificate.
-                new_tracker += 1;
-                continue;
-            }
+            for block_batch in blocks.chunks(CERTIFICATE_BATCH_SIZE) {
+                let first_block = block_batch.first().expect("nonempty batch");
+                let batch_size = block_batch.len() as u64;
+                let block_batch_range = BlockHeightRange::multi(first_block.height, batch_size);
+                let query = ChainInfoQuery::new(chain_id)
+                    .with_sent_certificate_hashes_in_range(block_batch_range.clone());
+
+                let local_response = self.local_query(query.clone()).await?;
+                let highest_found = Self::highest_found(&query, &local_response);
+                let blocks_missing = block_batch_range
+                    .highest_expected()
+                    .saturating_sub(highest_found)
+                    .0;
+
+                new_tracker += batch_size - blocks_missing;
```

Comment:
But these may not be in the same order as in the validator's `received_log`. We need to make sure we don't increment the tracker beyond the first gap.

Reply:
Moving changes to this method to a separate PR.